### PR TITLE
fix: cloning group fields

### LINKED ISSUE #172: Cloning a field group within a flexible content field without selecting "prefix field names"
### Description

When an ACF Field Group is cloned in a flexible content layout, the Schema fails to load if "prefix field names" is not selected.

### Steps to reproduce

 In a "vanilla" WordPress install with ACF Pro, WPGraphQL and WPGraphQL for ACF:

1. Import the field groups provided in a .zip below
2. Open GraphiQL
3. See that the Schema Loads
4. Search "AcfFieldGroup" in the Schema docs and see "PrimaryFieldGroup" and "SecondaryFieldGroup" as groups that implement the interface

CleanShot 2024-02-14 at 13 16 17

5. In the "Cloned Group" field within the "Primary Field Group" flexible content layout, disable the "Prefix Field Names" setting and save. 
6. Open GraphiQL again
7. See that the Schema no longer loads

CleanShot 2024-02-14 at 13 14 25

### PHP or JSON export of the ACF Field Group(s)

[acf-export-2024-02-14.json.zip](https://github.com/wp-graphql/wpgraphql-acf/files/14285677/acf-export-2024-02-14.json.zip)

### Additional context

_No response_

### WPGraphQL Version

1.21.0

### WPGraphQL For ACF Version

2.1.1

### ACF (Advanced Custom Fields) Version. Free or Pro?

6.2.2 PRO

### WordPress Version

6.4.3

### PHP Version

8.2

### Additional enviornment details

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have disabled ALL plugins except for WPGraphQL, WPGraphQL For ACF, ACF, etc.

- [X] Yes
- [ ] My issue is with compatibility with a specific WordPress plugin, and I have listed all my installed plugins (and version info) above.

### LINKED ISSUE #197: Schema fails to load when field group is cloned on many different flexible content layouts
### Description

When an ACF Field Group is cloned onto many different ACF Field Groups, the schema fails to load.

### Steps to reproduce

1. With the following plugins active: 

- WPGraphQL v1.23.0
- ACF PRO v6.2.7
- WPGraphQL for ACF v2.2.0
- WPGraphQL IDE 1.1.0 (optional)

2. Import the following ACF Field Groups export: [dev-site-acf-backup2.json](https://github.com/wp-graphql/wpgraphql-acf/files/15031238/dev-site-acf-backup2.json)

3. Try to load the GraphiQL IDE

**Expectation:** 
IDE loads and I can browse the schema and test queries, etc

**Actual:** 
IDE fails to load the schema and outputs the following errors: 

CleanShot 2024-04-18 at 16 50 35

CleanShot 2024-04-18 at 16 50 48

### PHP or JSON export of the ACF Field Group(s)

[dev-site-acf-backup2.json](https://github.com/wp-graphql/wpgraphql-acf/files/15031238/dev-site-acf-backup2.json) (also referenced above)

### Additional context

This appears to maybe be caused by the `array_merge_recursive()` within WPObjectType and WPInterfaceType that implements fields from Interfaces onto implementing Interfaces and Object types. 

Additionally, there appears to be some recursive functions that could bail early within WPGraphQL for ACF. 

The issue seems closely related to: 

- Group field types
- Clone field types
- Interface fields being applied to other Interfaces / Object Types

----

related:

- #193 
- https://github.com/wp-graphql/wp-graphql/pull/3100

### WPGraphQL Version

1.23.0

### WPGraphQL For ACF Version

2.22.0

### ACF (Advanced Custom Fields) Version. Free or Pro?

6.2.7 (pro)

### WordPress Version

6.5.2

### PHP Version

8.2

### Additional enviornment details

_No response_

### Please confirm that you have searched existing issues in the repo.

- [X] Yes

### Please confirm that you have disabled ALL plugins except for WPGraphQL, WPGraphQL For ACF, ACF, etc.

- [X] Yes
- [ ] My issue is with compatibility with a specific WordPress plugin, and I have listed all my installed plugins (and version info) above.

### PR DESCRIPTION
**Pull request by jasonbahl**
What does this implement/fix? Explain your changes.
---------------------------------------------------
This is an attempt at fixing a bug related to cloning field groups that have a field of the "group" field within them. 

For the situation described in #172 this gets the Schema to load and lets the fields be queried, but there are some things missing still. 

For example: the Interface for the field group being cloned isn't being applied properly to the Type representing the Flexible Content Layout

Does this close any currently open issues?
------------------------------------------

closes #172 
closes #197 

related: https://github.com/wp-graphql/wp-graphql/pull/3100

Any relevant logs, error output, GraphiQL screenshots, etc?
-------------------------------------

Given the ACF Field Groups in #172: 

### BEFORE:

Errors fetching the schema, as described in #172 

CleanShot 2024-04-05 at 15 33 34

### AFTER:

The Schema loads, and I can query data

CleanShot 2024-04-05 at 15 32 04

Any other comments?
-------------------
I _should_ be able to use the following fragment: 

```graphql
fragment ContentBlocksBlocksAccordionLayout on ContentBlocksBlocksAccordionLayout {
  ...BlockAccordion_Fields
}

fragment BlockAccordion_Fields on BlockAccordion_Fields {
    accordionItems {
    title
    description
  }
  description
  icon {
    node {
      mediaItemUrl
    }
  }
  sectionId
  title
  variant
}
```

~But this is showing invalid because `BlockAccordion_Fields` is being applied to `ContentBlocksBlocksAccordionLayoutAccordion` and not `ContentBlocksBlocksAccordionLayout`~

~I believe there's some correlation with how Flex Field Layouts are mapped to the schema in `/src/FieldType/FlexibleContent.php` and how cloned groups apply their interfaces in `src/FieldType/CloneField.php`, but I've not been able to pin it down quite yet.~

These fragments now work as expected.

### CODE REVIEW

**Comment by nhankla-kl** [comment]
Having some build errors but will log what we did to fix them.

**Review by josephfusco — DISMISSED** [review]
I was able to reproduce the original issue and confirm that the schema does not break with the changes in this PR 🙌🏻

**Comment by nhankla-kl** [comment]
Gettin Schema issues still:


Failure running getIntrospectionQuery: {
  "graphQLErrors": [
    {
      "message": "The field &#039;subLayout&#039; on Type &#039;PostSectionsPostSectionsContentLayout_Fields&#039; is configured to return &#039;PostContentSubLayout&#039; which is a non-existent Type in the Schema. Make sure to define a valid type for all fields. This might occur if there was a typo with &#039;PostContentSubLayout&#039;, or it needs to be registered to the Schema.",
      "extensions": {
        "category": "graphql"
      },
      "locations": [
        {
          "line": 15,
          "column": 5
        }
      ],
      "path": [
        "__schema",
        "types"
      ],
      "trace": [
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Type/Schema.php",
          "line": 406,
          "call": "WPGraphQL\\Registry\\TypeRegistry::WPGraphQL\\Registry\\{closure}()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Type/Definition/FieldDefinition.php",
          "line": 210,
          "call": "GraphQL\\Type\\Schema::resolveType()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Utils/TypeInfo.php",
          "line": 213,
          "call": "GraphQL\\Type\\Definition\\FieldDefinition::getType()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Type/Schema.php",
          "line": 222,
          "call": "GraphQL\\Utils\\TypeInfo::extractTypes()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Type/Schema.php",
          "line": 208,
          "call": "GraphQL\\Type\\Schema::collectAllTypes()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Type/Introspection.php",
          "line": 240,
          "call": "GraphQL\\Type\\Schema::getTypeMap()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 623,
          "call": "GraphQL\\Type\\Introspection::GraphQL\\Type\\{closure}()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 550,
          "call": "GraphQL\\Executor\\ReferenceExecutor::resolveFieldValueOrError()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 1195,
          "call": "GraphQL\\Executor\\ReferenceExecutor::resolveField()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 1145,
          "call": "GraphQL\\Executor\\ReferenceExecutor::executeFields()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 1106,
          "call": "GraphQL\\Executor\\ReferenceExecutor::collectAndExecuteSubfields()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 793,
          "call": "GraphQL\\Executor\\ReferenceExecutor::completeObjectValue()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 741,
          "call": "GraphQL\\Executor\\ReferenceExecutor::completeValue()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 654,
          "call": "GraphQL\\Executor\\ReferenceExecutor::completeValue()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 557,
          "call": "GraphQL\\Executor\\ReferenceExecutor::completeValueCatchingError()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 1195,
          "call": "GraphQL\\Executor\\ReferenceExecutor::resolveField()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 264,
          "call": "GraphQL\\Executor\\ReferenceExecutor::executeFields()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/ReferenceExecutor.php",
          "line": 215,
          "call": "GraphQL\\Executor\\ReferenceExecutor::executeOperation()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Executor/Executor.php",
          "line": 156,
          "call": "GraphQL\\Executor\\ReferenceExecutor::doExecute()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/GraphQL.php",
          "line": 162,
          "call": "GraphQL\\Executor\\Executor::promiseToExecute()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php",
          "line": 311,
          "call": "GraphQL\\GraphQL::promiseToExecute()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/Helper.php",
          "line": 211,
          "call": "GraphQL\\Server\\Helper::promiseToExecuteOperation()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/vendor/webonyx/graphql-php/src/Server/StandardServer.php",
          "line": 136,
          "call": "GraphQL\\Server\\Helper::executeOperation()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/src/Request.php",
          "line": 703,
          "call": "GraphQL\\Server\\StandardServer::executeRequest()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/src/Router.php",
          "line": 465,
          "call": "WPGraphQL\\Request::execute_http()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-content/plugins/wp-graphql/src/Router.php",
          "line": 257,
          "call": "WPGraphQL\\Router::process_http_request()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/class-wp-hook.php",
          "line": 324,
          "call": "WPGraphQL\\Router::resolve_http_request()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/class-wp-hook.php",
          "line": 348,
          "call": "WP_Hook::apply_filters()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/plugin.php",
          "line": 565,
          "call": "WP_Hook::do_action()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/class-wp.php",
          "line": 418,
          "function": "do_action_ref_array()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/class-wp.php",
          "line": 813,
          "call": "WP::parse_request()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-includes/functions.php",
          "line": 1336,
          "call": "WP::main()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/wp-blog-header.php",
          "line": 16,
          "function": "wp()"
        },
        {
          "file": "/sites/dev-back-vertice.kernandlead.com/files/index.php",
          "line": 17,
          "function": "require('/sites/dev-back-vertice.kernandlead.com/files/wp-blog-header.php')"
        }
      ]
    }
  ],
  "clientErrors": [],
  "networkError": null,
  "message": "The field &#039;subLayout&#039; on Type &#039;PostSectionsPostSectionsContentLayout_Fields&#039; is configured to return &#039;PostContentSubLayout&#039; which is a non-existent Type in the Schema. Make sure to define a valid type for all fields. This might occur if there was a typo with &#039;PostContentSubLayout&#039;, or it needs to be registered to the Schema."
}

**Comment by jasonbahl** [comment]
oops. didn't mean to close.

**Comment by jasonbahl** [comment]
@nhankla-kl was your ACF Field Groups working in the Schema prior to this PR and breaking in this PR?

**Comment by nhankla-kl** [comment]
@jasonbahl  it works in .5.5.3

**Comment by jasonbahl** [comment]
@nhankla-kl WPGraphQL for ACF v0.5.3? As in this version: https://github.com/wp-graphql/wp-graphql-acf/releases/tag/v0.5.3?

**Comment by nhankla-kl** [comment]
@jasonbahl  correct

Screenshot 2024-04-10 at 12 24 03 PM

**Comment by jasonbahl** [comment]
Ok so WPGraphQL for ACF v2.2.0 release also does not work, I assume? https://github.com/wp-graphql/wpgraphql-acf/releases/tag/v2.2.0

This specific PR did not break anything? 

If you are looking to migrate from v0.*.* to v2.*, I recommend checking the upgrade guide here: https://acf.wpgraphql.com/upgrade-guide/

It has some helpful information on things to look for in the Schema, such as avoiding special characters in GraphQL Type and Field names, avoiding field/type names starting with a number, etc. 

I did import your ACF field group and will see if I can help identify how to best support it, but it doesn't seem that this specific PR has caused any new issues compared to v2.2.0

**Comment by jasonbahl** [comment]
@nhankla-kl One thing I noticed is that you had a couple instances of different ACF Field Groups with the same GraphQL Type Name: 

- [module] Blog by Categories
- [module] Blog List

Both had the GraphQL Type Name: `ModuleBlog`

- [module] Content Grid
- [module] News

Both had the GraphQL Type Name `ModuleNews`

- [module] Video Hero
- [module] Hero

Both had the GraphQL Type Name `ModuleHero`

I'm still looking through things, but these are some observations so far.

**Comment by jasonbahl** [comment]
@nhankla-kl Ok, I think I'm getting close to understanding another bug. 

It appears that there is a bug when it comes to cloning a field group that contains other cloned field groups. 

For example, you have the `[module] Alternating Row` which has the following clone fields: 

- header
- media
- background
- slant
- border
- layout

It appears that somehow the double cloning isn't properly assigning GraphQL Interfaces to the schema. 

I'm working to better understand why this is and figure out a fix for it.

**Comment by jasonbahl** [comment]
@nhankla-kl I think I might have discovered the problem and am working on updating this PR with a solution. Hopefully will have something to test shortly 🤞🏻

**Comment by jasonbahl** [comment]
@nhankla-kl I've pushed up several changes to this PR and I believe we're getting close to having things work for the original scenario and the scenario you brought up if you want to give it another try.

File: docker/app.setup.sh [hunk 1]
```diff
@@ -36,15 +38,31 @@ apt-get -y update
...
+	if [[ -n ${WPGRAPHQL_GIT_REPO} ]]; then
+		echo "Installing WPGraphQL from GitHub repo ${WPGRAPHQL_GIT_REPO}"
+		# Clone the repository
+		git clone -b ${WPGRAPHQL_GIT_BRANCH} ${WPGRAPHQL_GIT_REPO} "${PLUGINS_DIR}/wp-graphql"
+		# Navigate to the plugin directory
+		cd "${PLUGINS_DIR}/wp-graphql"
+		# Install dependencies with Composer
+		composer install --no-dev
+		# Optionally activate the plugin using wp-cli
+		wp plugin activate wp-graphql --allow-root
```
Comment (jasonbahl) [inline docker/app.setup.sh:53]: NOTE: 

This allows us to run tests against arbitrary branches of WPGraphQL. 

For example, this feature relies on some work in this PR (https://github.com/wp-graphql/wp-graphql/pull/3100)

So, I was able to update my `.env.testing` file with these values: 

```
 WPGRAPHQL_GIT_REPO="https://github.com/jasonbahl/wp-graphql.git"
 WPGRAPHQL_GIT_BRANCH="fix/interface-recursion"
```

And tests pass when running against that branch of WPGraphQL 👏🏻 👏🏻 

CleanShot 2024-04-19 at 11 40 22

**Comment by jasonbahl** [comment]
@nhankla-kl I've updated #172 with a comment outlining steps to reproduce using the exported ACF Field Group you have provided. 

Using your ACF Field Group, I am now able to load the Schema. 

### Before

Schema doesn't load properly and errors out the GraphiQL IDE

CleanShot 2024-04-22 at 10 07 50


### After 

Schema loads and I can use the GraphiQL IDE:

CleanShot 2024-04-22 at 10 08 22

----

Test added using the exported ACF Field Group provided above as an imported field group with assertions on querying the possible types and on executing an example query: https://github.com/wp-graphql/wpgraphql-acf/pull/193/commits/43d2b91938bb24ff92a0895da5f504965ed5f712

**Comment by nhankla-kl** [comment]
@jasonbahl  this is amazing will test it on several dev sites that have similar ACF structure. Will let you know. Thanks!

**Comment by jasonbahl** [comment]
related: https://github.com/wp-graphql/wp-graphql/pull/3102 (specifically, see: https://github.com/wp-graphql/wp-graphql/pull/3100)

**Comment by jasonbahl** [comment]
FWIW, I'm currently chasing down one last bug. 

Basically, when I get things to work for this test: (`TestCloneGroupWithoutPrefixCest`) https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1932

These tests fail: 

- `TestCloneFieldsCest`: https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1930
- `TestCloneWithGroupCest`: https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1934
- `TestCloneWithRepeaterCest`: https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1935
- `TestCloneWithRepeaterCest`:https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1936
- `TestCloneWithRepeaterCest`:https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8806878088/job/24172651081?pr=193#step:7:1937

And vis-versa. 

So I still have a bit of digging and conflict resolution. 🤔

**Comment by jasonbahl** [comment]
And here we are, with all tests passing! 🥳 

https://github.com/wp-graphql/wpgraphql-acf/actions/runs/8810691564

**Comment by jasonbahl** [comment]
@nhankla-kl fwiw, it seems like there still are some more bugs related to cloning field groups that have nested group fields. 

I've broken it down into another issue that I'm continuing to troubleshoot: https://github.com/wp-graphql/wpgraphql-acf/issues/201

**Pull request merged**